logstash-core/src/main/java/org/logstash/config/ir/graph/Graph.java

```diff
@@ -0,0 +1,434 @@
+        public final Map<Vertex, Vertex> oldToNewVertices;
+        public final Map<Edge, Edge> oldToNewEdges;
+
+        GraphCombinationResult(Graph graph, Map<Vertex, Vertex> oldToNewVertices, Map<Edge, Edge> oldToNewEdges) {
+            this.graph = graph;
+            this.oldToNewVertices = oldToNewVertices;
+            this.oldToNewEdges = oldToNewEdges;
+        }
+    }
+
+    /*
+      Return a copy of this graph with the other graph's nodes to this one by connection this graph's leaves to
+      the other graph's root
+    */
+    public Graph chain(Graph otherGraph) throws InvalidIRException {
+        if (otherGraph.getVertices().size() == 0) return this.copy();
+        if (this.isEmpty()) return otherGraph.copy();
+
+        GraphCombinationResult combineResult = Graph.combine(this, otherGraph);
+
+        // Build these lists here since we do mutate the graph in place later
+        // This isn't strictly necessary, but makes things less confusing
+        Collection<Vertex> fromLeaves = allLeaves().map(combineResult.oldToNewVertices::get).collect(Collectors.toSet());
+        Collection<Vertex> toRoots = otherGraph.roots().map(combineResult.oldToNewVertices::get).collect(Collectors.toSet());
+
+        return combineResult.graph.chain(fromLeaves, toRoots);
+    }
+
+    public Graph chain(Vertex... otherVertex) throws InvalidIRException {
+        chain(this.getAllLeaves(), Arrays.asList(otherVertex));
+        return this;
+    }
+
+    // This does *not* return a copy for performance reasons
+    private Graph chain(Collection<Vertex> fromLeaves, Collection<Vertex> toVertices) throws InvalidIRException {
+        for (Vertex leaf : fromLeaves) {
+            for (Edge.EdgeFactory unusedEf : leaf.getUnusedOutgoingEdgeFactories()) {
+                for (Vertex toVertex : toVertices) {
+                    this.threadVertices(unusedEf, leaf, toVertex);
+                }
+            }
+        }
+
+        return this;
+    }
+
+    public Collection<Edge> threadVerticesById(String... vertexIds) throws InvalidIRException {
+        return threadVerticesById(PlainEdge.factory, vertexIds);
+    }
+
+    public Collection<Edge> threadVerticesById(Edge.EdgeFactory edgeFactory, String... vertexIds) throws InvalidIRException {
+        Vertex[] argVertices = new Vertex[vertexIds.length];
+        for (int i = 0; i < vertexIds.length; i ++) {
+            String id = vertexIds[i];
+            Vertex v = getVertexById(id);
+            if (v==null) throw new InvalidIRException("Could not thread vertex, id not found in graph: !" + id + "\n" + this);
+            argVertices[i] = v;
+        }
+        return threadVertices(edgeFactory, argVertices);
+    }
+
+    // Will not validate the graph after running!
+    // You must invoke validate the graph yourself
+    // after invoking
+    public Collection<Edge> threadVerticesUnsafe(Edge.EdgeFactory edgeFactory, Vertex... argVertices) throws InvalidIRException {
```

Comment:
"thread" is an odd verb here. Is there etymology or a story to tell about the name?